process/block/preprocess/basePreProcess.go

```diff
@@ -218,17 +217,39 @@ func (bpp *basePreProcess) computeExistingAndMissing(
 		}
 
 		if len(txHashes) > 0 {
-			tmp := &txsHashesInfo{
-				txHashes:        sliceUtil.TrimSliceSliceByte(txHashes),
-				receiverShardID: miniBlock.ReceiverShardID,
-			}
-			missingTxsForShard[miniBlock.SenderShardID] = append(missingTxsForShard[miniBlock.SenderShardID], tmp)
+			bpp.setMissingTxsForShard(miniBlock.SenderShardID, miniBlock.ReceiverShardID, txHashes, forBlock)
+			missingTxsForShard[miniBlock.SenderShardID] = append(missingTxsForShard[miniBlock.SenderShardID], txHashes...)
 		}
+
 		txHashes = txHashes[:0]
 	}
-	forBlock.mutTxsForBlock.Unlock()
 
-	return missingTxsForShard
+	return bpp.requestMissingTxsForShard(missingTxsForShard, onRequestTxs)
+}
+
+func (bpp *basePreProcess) setMissingTxsForShard(
+	senderShardID uint32,
+	receiverShardID uint32,
+	txHashes [][]byte,
+	forBlock *txsForBlock,
+) {
+	txShardInfoToSet := &txShardInfo{senderShardID: senderShardID, receiverShardID: receiverShardID}
+	for _, txHash := range txHashes {
+		forBlock.txHashAndInfo[string(txHash)] = &txInfo{tx: nil, txShardInfo: txShardInfoToSet}
+	}
+}
+
+func (bpp *basePreProcess) requestMissingTxsForShard(
```

Comment:
maybe add comment as it should be called only under mutex

Reply:
added